// @google-cloud/logging/build/src/middleware/express/make-http-request.js
"use strict";
/*!
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeHttpRequestData = void 0;
function makeHttpRequestData(req, res, latencyMilliseconds) {
    return {
        status: res.statusCode,
        requestUrl: req.originalUrl,
        requestMethod: req.method,
        userAgent: req.headers['user-agent'],
        responseSize: (res.getHeader && Number(res.getHeader('Content-Length'))) || 0,
        latency: {
            seconds: Math.floor(latencyMilliseconds / 1e3),
            nanos: Math.floor((latencyMilliseconds % 1e3) * 1e6),
        },
    };
}
exports.makeHttpRequestData = makeHttpRequestData;
